src/bindings/python/flux/job.py

```diff
@@ -85,3 +96,385 @@ def wait(flux_handle, jobid=lib.FLUX_JOBID_ANY):
+    def duration(self, duration):
+        """
+        Assign a time limit to the job.  The duration may be:
+        - a float in seconds
+        - a string in Flux Standard Duration
+        A duration of zero is interpreted as "not set".
+        """
+        if isinstance(duration, six.string_types):
+            time = parse_fsd(duration)
+        elif isinstance(duration, float):
+            time = duration
+        else:
+            raise ValueError("duration must be a float or string")
+        if time < 0:
+            raise ValueError("duration must not be negative")
+        if math.isnan(time) or math.isinf(time):
+            raise ValueError("duration must be a normal, finite value")
+        self.jobspec["attributes"]["system"]["duration"] = time
+
+    @property
+    def cwd(self):
+        """
+        Get working directory of job.
+        """
+        try:
+            self.jobspec["attributes"]["system"]["cwd"]
+        except KeyError:
+            return
+
+    @cwd.setter
+    def cwd(self, cwd):
+        """
+        Set working directory of job.
+        """
+        if not isinstance(cwd, six.string_types):
+            raise ValueError("cwd must be a string")
+        self.jobspec["attributes"]["system"]["cwd"] = cwd
+
+    @property
+    def environment(self):
+        """
+        Get (entire) environment of job.
+        """
+        try:
+            self.jobspec["attributes"]["system"]["environment"]
+        except KeyError:
+            return
+
+    @environment.setter
+    def environment(self, environ):
+        """
+        Set (entire) environment of job.
+        """
+        if not isinstance(environ, collectionsAbc.Mapping):
+            raise ValueError("environment must be a mapping")
+        self.jobspec["attributes"]["system"]["environment"] = environ
+
+    def _set_treedict(self, in_dict, key, val):
+        """
+        _set_treedict(d, "a.b.c", 42) is like d[a][b][c] = 42
+        but levels are created on demand.
+        """
+        path = key.split(".", 1)
+        if len(path) == 2:
+            self._set_treedict(in_dict.setdefault(path[0], {}), path[1], val)
+        else:
+            in_dict[key] = val
+
+    def setattr(self, key, val):
+        """
+        set job attribute
+        """
+        self._set_treedict(self.jobspec, "attributes." + key, val)
+
+    def setattr_shopt(self, key, val):
+        """
+        set job attribute: shell option
+        """
+        self.setattr("system.shell.options." + key, val)
+
+    def dumps(self):
+        return json.dumps(self.jobspec)
+
+    @property
+    def resources(self):
+        return self.jobspec.get("resources", None)
+
+    @property
+    def tasks(self):
+        return self.jobspec.get("tasks", None)
+
+    @property
+    def attributes(self):
+        return self.jobspec.get("attributes", None)
+
+    @property
+    def version(self):
+        return self.jobspec.get("version", None)
+
+    def resource_walk(self):
```

Comment:
Does this match how the filesystem walk and kvs walk methods work? I can't remember off top of my head but it would be good if they work similarly if the name is basically the same.

Reply:
The `os.walk` and `kvs.walk` functions yields a tuple of `(root, dirs, files)` for each call to `next` while `resource_walk` returns just the resource.  I'm not sure the dir/file distinction makes sense here.  Those walks also support the `topdown` argument.    That one could very well make sense here.

I could also change this to a `__iter__` method to make the jobspec object an iterator.

Reply:
Ok, `__iter__` probably makes sense.  I agree that the dirs/files split doesn't make sense here, but `(parent, resource)` might?  Generally the `walk` functions are a more general kind of traversal than a linear order, and also I think do multi-visits in some cases, so differentiating it seems like it would be good.

Reply:
I switched this method to `__iter__` and then added a new `resource_walk` method that (pasted from docstring):

>         Performs a depth-first, pre-order traversal. Yields a tuple containing
>         (parent, resource, count).  `parent` is None when `resource` is a
>         top-level resource. `count` is the number of that resource including the
>         multiplicative effects of the `with` clause in ancestor resources.  For
>         example, the following resource section, will yield a count of 2 for the
>         `slot` and a count of 8 for the `core` resource.
> 
>         ```yaml
>         - type: slot
>           count: 2
>           with:
>             - type: core
>               count: 4
>         ```

Reply:
That seems pretty useful, thanks!